### PR TITLE
979 - Standardize Terraform Deployment From Local

### PR DESCRIPTION
**Pull request by jamesobrooks**
Grabs deployment-related environment variables from AWS Secrets Manager during `deploy-app.sh` for web-api, web-client, and environment-specific deployments. This makes it easy to deploy a portion of the app from the local machine without the risk of accidentally provisioning too many resources, for example two or more ES instances for environments where more than one is not needed. Examples shown below, but may be accessed [via AWS console](https://console.aws.amazon.com/secretsmanager/home?region=us-east-1#!/listSecrets/).

**exp2_deploy:**
```json
{
  "EFCMS_DOMAIN": "exp2.ustc-case-mgmt.flexion.us",
  "ZONE_NAME": "ustc-case-mgmt.flexion.us",
  "ENVIRONMENT": "exp2",
  "DYNAMSOFT_PRODUCT_KEYS": "noop",
  "DYNAMSOFT_S3_ZIP_PATH": "s3://ustc-case-mgmt.flexion.us-dynamsoft/dynamic-web-twain-sdk-14.3.1.tar.gz",
  "DYNAMSOFT_URL": "https://dynamsoft-lib-dev.ustc-case-mgmt.flexion.us",
  "IS_DYNAMSOFT_ENABLED": 0,
  "ENABLE_HEALTH_CHECKS": 0,
  "BOUNCED_EMAIL_RECIPIENT": "",
  "COGNITO_SUFFIX": "flexion-efcms",
  "DISABLE_EMAILS": true,
  "EMAIL_DMARC_POLICY": "v=DMARC1; p=none; rua=mailto:flexionustc@gmail.com;",
  "ES_INSTANCE_COUNT": 1,
  "ES_INSTANCE_TYPE": "t2.small.elasticsearch",
  "ES_VOLUME_SIZE": 10,
  "IRS_SUPERUSER_EMAIL": "flexionustc+irsSuperUser@gmail.com"
}
```

**account_deploy:**
```json
{
  "ZONE_NAME": "ustc-case-mgmt.flexion.us",
  "ES_LOGS_INSTANCE_COUNT": 1,
  "ES_LOGS_INSTANCE_TYPE": "t2.medium.elasticsearch",
  "ES_LOGS_EBS_VOLUME_SIZE_GB": 20,
  "COGNITO_SUFFIX": "flexion-efcms"
}
```

**When this goes to the court environments, secrets will need to have been created, and the IAM roles updated to grant CircleCI access to the secrets.**

### CODE REVIEW

File: web-client/terraform/bin/deploy-app.sh [hunk 1]
```diff
@@ -2,17 +2,37 @@
 
 ENVIRONMENT=$1
 
+REGION=us-east-1
+content=$(aws secretsmanager get-secret-value --region ${REGION} --secret-id "exp2_deploy" --query "SecretString" --output text)
```
Comment (codyseibert) [inline web-client/terraform/bin/deploy-app.sh]: --secret-id "exp2_deploy" needs to be changed to "${ENVIRONMENT}_deploy"

File: web-client/terraform/bin/deploy-app.sh [hunk 2]
```diff
@@ -2,17 +2,37 @@
...
+echo ${content} | jq -r 'to_entries|map("\(.key)=\"\(.value)\"")|.[]' > .env
+set -o allexport
+source .env
+set +o allexport
+
+[ -z "${DYNAMSOFT_PRODUCT_KEYS}" ] && echo "You must have DYNAMSOFT_PRODUCT_KEYS set in your environment" && exit 1
+[ -z "${DYNAMSOFT_S3_ZIP_PATH}" ] && echo "You must have DYNAMSOFT_S3_ZIP_PATH set in your environment" && exit 1
+[ -z "${DYNAMSOFT_URL}" ] && echo "You must have DYNAMSOFT_URL set in your environment" && exit 1
```
Comment (codyseibert) [inline web-client/terraform/bin/deploy-app.sh:14]: 👍  thanks for adding these extra checks in.  I didn't even know they were missing

**Review by codyseibert — CHANGES_REQUESTED** [review]

File: web-client/terraform/bin/deploy-app.sh (hunk 1 above)
Comment (jamesobrooks) [reply web-client/terraform/bin/deploy-app.sh]:
```suggestion
content=$(aws secretsmanager get-secret-value --region ${REGION} --secret-id "${ENVIRONMENT}_deploy" --query "SecretString" --output text)
```

Comment (jamesobrooks) [reply web-client/terraform/bin/deploy-app.sh]: Will do.

**Review by codyseibert — APPROVED** [review]

File: docs/environments.md [hunk 3]
```diff
@@ -61,29 +60,28 @@ A prerequisite for a successful build within CircleCI is [access to CircleCI’s
...
+  | `EMAIL_DMARC_POLICY`*       | DMARC policy in the format of `v=DMARC1; p=none; rua=mailto:postmaster@example.com;` |
+  | `IRS_SUPERUSER_EMAIL`*      | Email address used to serve all new petitions to the IRS |
+  | `DEFAULT_ACCOUNT_PASS`      | Default password for all test accounts and some password resets |
+  | `STATUSPAGE_DNS_RECORD`     | DNS record for Statuspage of CNAME `status.${EFCMS_DOMAIN}` (optional) |
+  | `SESSION_MODAL_TIMEOUT`     | Time in ms to wait before logging the user out after the idle timeout modal displays (optional, default: `300000` / 5 mins) |
+  | `SESSION_TIMEOUT`           | Time in ms to wait displaying the idle timeout modal (optional, default: `3300000` / 55 mins) |
+  | `CLIENT_STAGE`              | The `process.env.STAGE` for the React application |
+  | `BOUNCED_EMAIL_RECIPIENT`*  | An email to which email bounced should be sent (defaults to noreply@`EFCMS_DOMAIN`) |
+  | `PROD_ENV_ACCOUNT_ID`       | The account ID of the AWS account with Production Data |
+  | `LOWER_ENV_ACCOUNT_ID`      | The account ID of the AWS account where copies of Production Data might live |
```
Comment (jamesobrooks) [inline docs/environments.md]:
```suggestion
  | `AWS_ACCOUNT_ID` | AWS account ID without hyphens |
  | `AWS_ACCESS_KEY_ID` | AWS access key for the AWS CircleCI user |
  | `AWS_SECRET_ACCESS_KEY` | AWS secret access key for the AWS CircleCI user |
  | `DYNAMSOFT_PRODUCT_KEYS`* | Dynamsoft Web TWAIN product key used |
  | `DYNAMSOFT_S3_ZIP_PATH`* | Dynamsoft Web TWAIN full S3 path ZIP configured above, e.g. `s3://bucketname/Dynamsoft/dynamic-web-twain-sdk-14.3.1.tar.gz` |
  | `EFCMS_DOMAIN`* | Domain name chosen above |
  | `COGNITO_SUFFIX`* | Suffix of your choice for the Cognito URL |
  | `USTC_ADMIN_USER` | Username of your choice used by the Cognito admin user |
  | `USTC_ADMIN_PASS` | Password of your choice used by the Cognito admin user |
  | `EMAIL_DMARC_POLICY`* | DMARC policy in the format of `v=DMARC1; p=none; rua=mailto:postmaster@example.com;` |
  | `IRS_SUPERUSER_EMAIL`* | Email address used to serve all new petitions to the IRS |
  | `DEFAULT_ACCOUNT_PASS` | Default password for all test accounts and some password resets |
  | `STATUSPAGE_DNS_RECORD` | DNS record for Statuspage of CNAME `status.${EFCMS_DOMAIN}` (optional) |
  | `SESSION_MODAL_TIMEOUT` | Time in ms to wait before logging the user out after the idle timeout modal displays (optional, default: `300000` / 5 mins) |
  | `SESSION_TIMEOUT` | Time in ms to wait displaying the idle timeout modal (optional, default: `3300000` / 55 mins) |
  | `CLIENT_STAGE` | The `process.env.STAGE` for the React application |
  | `BOUNCED_EMAIL_RECIPIENT`* | An email to which email bounced should be sent (defaults to noreply@`EFCMS_DOMAIN`) |
  | `PROD_ENV_ACCOUNT_ID` | The account ID of the AWS account with Production Data |
  | `LOWER_ENV_ACCOUNT_ID` | The account ID of the AWS account where copies of Production Data might live |
```

**Review by rachelschneiderman — APPROVED** [review]

File: iam/terraform/account-specific/bin/deploy-app.sh [hunk 4]
```diff
@@ -1,5 +1,12 @@
 #!/bin/bash -e
 
+REGION=us-east-1
+content=$(aws secretsmanager get-secret-value --region ${REGION} --secret-id "account_deploy" --query "SecretString" --output text)
+echo ${content} | jq -r 'to_entries|map("\(.key)=\"\(.value)\"")|.[]' > .env
+set -o allexport
+source .env
+set +o allexport
```
Comment (halprin) [inline iam/terraform/account-specific/bin/deploy-app.sh:8]: Can we add a comment to the top of this set of code explaining what is is doing?  E.g. something like...
`#Getting the account-wide deployment settings and injecting them into the shell environment`

File: iam/terraform/account-specific/main/circle-ci.tf [hunk 5]
```diff
@@ -158,7 +158,12 @@ resource "aws_iam_policy" "circle_ci_policy" {
         "s3:*",
         "cloudformation:*",
         "cloudwatch:*",
-        "lambda:*"
+        "lambda:*",
+        "secretsmanager:GetSecretValue",
+        "secretsmanager:GetResourcePolicy",
+        "secretsmanager:DescribeSecret",
+        "secretsmanager:GetRandomPassword",
+        "secretsmanager:ListSecretVersionIds"
```
Comment (halprin) [inline iam/terraform/account-specific/main/circle-ci.tf:166]: I would recommend that we move these actions into a separate section who's resource isn't `*` but specifically lists `account_deploy` and the environments secret.

File: web-api/terraform/bin/deploy-app.sh [hunk 6]
```diff
@@ -2,7 +2,16 @@
...
-[ -z "${CIRCLE_BRANCH}" ] && echo "You must have CIRCLE_BRANCH set in your environment" && exit 1
+export DEPLOYING_COLOR=(sh ./scripts/get-deploying-color.sh ${ENVIRONMENT})
+export MIGRATE_FLAG=(sh ./scripts/get-migrate-flag.sh ${ENVIRONMENT})
+
+REGION=us-east-1
+content=$(aws secretsmanager get-secret-value --region ${REGION} --secret-id "${ENVIRONMENT}_deploy" --query "SecretString" --output text)
+echo ${content} | jq -r 'to_entries|map("\(.key)=\"\(.value)\"")|.[]' > .env
+set -o allexport
+source .env
+set +o allexport
```
Comment (halprin) [inline web-api/terraform/bin/deploy-app.sh:13]: I'd like to see a comment here like I mentioned above.

File: web-client/terraform/bin/deploy-app.sh [hunk 7]
```diff
@@ -2,17 +2,37 @@
 
 ENVIRONMENT=$1
 
+REGION=us-east-1
+content=$(aws secretsmanager get-secret-value --region ${REGION} --secret-id "${ENVIRONMENT}_deploy" --query "SecretString" --output text)
+echo ${content} | jq -r 'to_entries|map("\(.key)=\"\(.value)\"")|.[]' > .env
+set -o allexport
+source .env
+set +o allexport
```
Comment (halprin) [inline web-client/terraform/bin/deploy-app.sh:10]: Comment, blah blah blah.

**Review by halprin — COMMENTED** [review]
This is super cool!

**Pull request merged**